### ISSUE TITLE
StringValidator crashes if \Yii::$app is not set.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#13254]
### What steps will reproduce the problem?
Create a `StringValidator` instance without having `Yii::$app` set.

### What is the expected result?
String validation should work fine without an app.

### What do you get instead?
Exception `trying to get property of non-object`.

### Additional info
The cause is:
````
if ($this->encoding === null) {
    $this->encoding = Yii::$app->charset;
}
````

I think it would be nice if there was a check to see if `Yii::$app` is set.

| Q                | A
| ---------------- | ---
| Yii version      | latest
| PHP version      | N/A
| Operating system | N/A

**Comment by SamMousa** [comment]
Same applies to `NumberValidator`:
````
$options = [
            'pattern' => new JsExpression($this->integerOnly ? $this->integerPattern : $this->numberPattern),
            'message' => Yii::$app->getI18n()->format($this->message, [
                'attribute' => $label,
            ], Yii::$app->language),
        ];
````

**Comment by klimov-paul** [comment]
This is questionable issue.
There are many places around validators and helpers, which depend on application instance.
For example:

- `StringHelper`: https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseStringHelper.php#L108 , https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseStringHelper.php#L111 , https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseStringHelper.php#L155 , https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseStringHelper.php#L212, https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseStringHelper.php#L237
- ArrayHelper: https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseArrayHelper.php#L607
- FileHelper: https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseFileHelper.php#L96

The most obvious one is `yii\helpers\Url`: https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseUrl.php#L136, https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseUrl.php#L142, https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseUrl.php#L219 and so on.

Yii architecture has a relatively high coupling between different classes - it is a price for the simpicity. I can see no way we can make each component trully independent without changing the main paradigm.

For me it is expected that I need Yii application instantiated before using most of the functionality.

**Comment by SamMousa** [comment]
@klimov-paul I agree, it is definitely not a "high priority" issue.
But less coupling is something to strive for in my opinion.

That said:
- I fully understand if the team doesn't want to spend time on this.
- I am still wondering if you would accept pull requests that change validators to do a "check before use" on `\Yii::$app`.

**Comment by cebe** [comment]
@klimov-paul I most of the places where the dependecy to `Yii::$app` is only for the charset we have a check already. There are only a few places like this one that are not covered. We can not and do not want to achive full independence but most components (if specially configured) can work without `\Yii::$app` without problem. E.g. a DbCache reads its db from Yii::$app->db but if the DB is explicitly configured, the component can be used independent. For the charset which is hardly ever different from UTF-8 a fallback is useful imo.

**Comment by klimov-paul** [comment]
> In most of the places where the dependecy to Yii::$app is only for the charset we have a check already.

Here is a `Application::$language` dependency:
https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseFileHelper.php#L96

> For the charset which is hardly ever different from UTF-8 a fallback is useful imo.

For me the issue does not worth the effots, but I will not object if you wish to provide fallbacks for `Application::$charset` and `Application::$language`. However in this case fallback should cover every usage not just some of them.

**Comment by SamMousa** [comment]
@klimov-paul With every usage I hope you mean every usage inside a single class not inside all framework classes?

**Comment by SamMousa** [comment]
I have implemented a solution for all validators:
- All validators used to call `Yii::$app->getI18n()->format($message, $params, Yii::$app->language)`
- I introduced `Validator::format($message, $params)`, which forwards the call if `Yii::$app exists` and otherwise uses a simplified implementation (same as `Yii::t` currently does).

This reduces the link to `Yii::$app` only 1 place in the base validator instead of repeated in each call to `format`. It also makes the dependency optional.
